Feat/issue 21564/autosync toggle action

ebracha opened the pull request:
<!--
Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.
-->

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [ ] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

<!-- Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request. -->

agaudreault commented:
This would be a breaking change... What about only adding a toggle auto-sync action on the application, but not changing the logic of the internal button?

crenshaw-dev commented:
Would be nice to have consistent permissioning in the UI... Maybe the button could try the new API first and fall back. The backend could log a deprecation warning for the old API (assuming it's its own endpoint).

agaudreault commented:
I think the approach @crenshaw-dev suggested has value because it allows to test the breaking changes and to opt-in the feature. 

You can do 2 separate PRs so it is easier to review. Implement approach 1, merge, then implement approach 2.

I would add a requirement to 2)
- Add a feature flag to enable the use of actions rbac instead of update. I think this should be opt-in because multiple people might have given permissions on `action/*` already. If we default (or fallback) to action/* without a flag, then we might cause users to have more permissions than what the admins expected.

ebracha commented:
yea good point doing it as an opt-in feature. I've modified this PR to follow the first approach:

1. Added the `toggle-auto-sync` action as a separate capability
2. Kept the existing UI button behavior unchanged
3. Maintained backward compatibility

Once this PR is merged, I'll create a new PR for the second approach that will:
1. Add a feature flag for action-based auto-sync (opt-in)
2. Modify the UI to try the action first and fall back to update permission
3. Add deprecation warnings for the update-based approach

Since the UI changes involve feature flags and deprecation warnings, should I create a new issue or link to #21564?

agaudreault commented on resource_customizations/argoproj.io/Application/health.lua:
The health of an Applciation should be removed from this PR.

agaudreault commented on resource_customizations/argoproj.io/Application/actions/toggle-auto-sync/action.lua:
```diff
@@ -0,0 +1,18 @@
+function toggleAutoSync(obj)
+    if obj.spec.syncPolicy and obj.spec.syncPolicy.automated then
+        obj.spec.syncPolicy.automated = nil
```
Missing a check to set `obj.spec.syncPolicy.automated.enabled = false` if it is currently true

agaudreault commented on resource_customizations/argoproj.io/Application/actions/actions_test.yaml:
Add tests for 
1) syncPolicy is null and toggle is performed => should have enabled true in spec
2) `syncPolicy.automated: {}` is defined and toggle is performed => should have enabled false in spec
